### PR TITLE
Fix for multi-GPU scattering

### PR DESCRIPTION
**Pull request by anakin-datawalker**
This is a fix for multi-GPU scattering issues as previously raised. It was tested on torch 1.7, cuda 10.2 and 11.0

### CODE REVIEW

**Comment by edouardoyallon** [comment]
The error message is slightly weird ( https://jenkins.flatironinstitute.org/job/kymatio/job/PR-640/1/console ) but as travis is working, I guess it's mainly linked to CUDA.

**Comment by janden** [comment]
Looks like it has to do with dependencies not installing properly. Since we haven't run the PR pipeline for a while, it's likely that some packages have been updated in the meantime. Will take a closer look.

File: kymatio/scattering2d/backend/torch_skcuda_backend.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
 from ...backend.torch_backend import _is_complex
 from ...backend.torch_skcuda_backend import cdgmm
 
-@cupy.util.memoize(for_each_device=True)
+@cupy._util.memoize(for_each_device=True)
```
Comment (janden) [inline kymatio/scattering2d/backend/torch_skcuda_backend.py:14]: (Next time, it's probably better to put this in a separate PR, but this is good for now.)

**Review by janden — APPROVED** [review]
This looks good. Just need to sort out the testing issues and see if it passes.

**Comment by janden** [comment]
Note: Changed merge branch to dev.

**Comment by janden** [comment]
Once #641 is merged, we'll rebase this one on top and try again. That should make the Jenkins issue go away.

**Comment by janden** [comment]
Rebased this onto current dev and reworded the commit message. Let's see how this does.

**Comment by janden** [comment]
Looks like it's going through. @anakin-datawalker can you provide a short snippet that demonstrates multi-GPU usage?

**Comment by anakin-datawalker** [comment]
Sure. I had provided the following code in the original issue, which raises the error "TypeError: Input and filter must be on the same GPU." when executed on current kymatio master but which runs correctly with the proposed PR. Would that be sufficient ?

```
import torch
from kymatio.torch import Scattering2D
scattering = Scattering2D(J=2, shape=(32, 32), max_order=1)
scattering = torch.nn.DataParallel(scattering).cuda()
x = torch.rand(8,8,32,32).cuda()
scattering(x)
```
Best

**Comment by janden** [comment]
That should be sufficient. Let me try it out. What did you say the minimum version of Torch was for which FFT supports DataParallel?

**Comment by anakin-datawalker** [comment]
I am almost sure it was torch 1.5.

**Comment by eickenberg** [comment]
This change looks very good to me, but I am having slight troubles fully understanding what is going on.

What I would expect is that there was an error of the type that we had filters registered as buffers, but we collected them in a dictionary for ease of manipulation. Then when it came to using built-in pytorch mechanisms to transport to different GPUs, e.g. `DataParallel`, the dictionary did not get updated and its contents still pointed to the tensors on the original GPU.

I see that one correction is not to write the registered buffer to the dictionary, and probably to create the dictionary when requested, but I can't connect all the dots here from looking at the code.
Is my assessment from above correct, or is there something else at play? If my assessment is correct, then how have we survived the equivalent problem of going from cpu to gpu?

**Comment by edouardoyallon** [comment]
> This change looks very good to me, but I am having slight troubles fully understanding what is going on.
> 
> What I would expect is that there was an error of the type that we had filters registered as buffers, but we collected them in a dictionary for ease of manipulation. Then when it came to using built-in pytorch mechanisms to transport to different GPUs, e.g. `DataParallel`, the dictionary did not get updated and its contents still pointed to the tensors on the original GPU.
> 
> I see that one correction is not to write the registered buffer to the dictionary, and probably to create the dictionary when requested, but I can't connect all the dots here from looking at the code.
> Is my assessment from above correct, or is there something else at play? If my assessment is correct, then how have we survived the equivalent problem of going from cpu to gpu?

Me neither but I guess wes hould acknowledge both pytorch fix and @anakin-datawalker 's work. If they're using this on their cluster, since it at least makes the code slightly more clean, then I'm in favor to merge it.
